[ews-build.webkit.org] Use twisted deferred to communicate with commits.webkit.org

JonWBedard opened the pull request:
#### eb4a573dc73211ab84be6af794e2fd01508d4ed6
<pre>
[ews-build.webkit.org] Use twisted deferred to communicate with commits.webkit.org
<a href="https://bugs.webkit.org/show_bug.cgi?id=249174">https://bugs.webkit.org/show_bug.cgi?id=249174</a>
rdar://103269096

Reviewed by Dewei Zhu and Aakash Jain.

* Tools/CISupport/ews-build/steps.py:
(DetermineLandedIdentifier): Inherit from ShellCommandNewStyle.
(DetermineLandedIdentifier.__init__): Ditto.
(DetermineLandedIdentifier.run): Replace start and evaluateCommand.
(DetermineLandedIdentifier.identifier_for_hash): Use TwistedAdditions.request and return
a deferred object.
(DetermineLandedIdentifier.start): Deleted.
(DetermineLandedIdentifier.url_for_hash_details): Deleted.
(DetermineLandedIdentifier.evaluateCommand): Deleted.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/258032@main">https://commits.webkit.org/258032@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/949795236e7b26b1378d764d390f01b524c747c7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/100716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9860 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/33757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/110015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/104704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/10800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/93117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/107861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106499 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/10800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/33757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/93117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/10800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/33757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/93117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/3554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/33757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/3578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/99644 "Passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/9689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/33757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/5358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2881 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -4905,25 +4977,30 @@ def evaluateCommand(self, cmd):
+            response = yield TwistedAdditions.request(
+                url=f'{COMMITS_INFO_URL}{hash}/json', logger=print,
+            )
+            if response.status_code == 200:
```
can response be None? if so we `response.status_code` will raise an exception. We should safeguard against that.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -4905,25 +4977,30 @@ def evaluateCommand(self, cmd):
+                url=f'{COMMITS_INFO_URL}{hash}/json', logger=print,
+            )
+            if response.status_code == 200:
+                defer.returnValue(response.json().get('identifier', '{}'.format(hash)).replace('@trunk', '@main'))
```
can use f-string instead of .format()

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -4905,25 +4977,30 @@ def evaluateCommand(self, cmd):
-            print(e)
-        return hash
+                defer.returnValue(hash)
+        except json.decoder.JSONDecodeError:
```
we are replacing generic catch-all with only JSONDecodeError. Just wanted to point-out to ensure this is intentional.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -67,6 +72,74 @@
+        logger = logger or (lambda _: None)
+        typ = type or b'GET'
+
+        if not url:
```
since `url` is mandatory, let's not make it a optional parameter in `request(url=None)`

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -67,6 +72,74 @@
+                url=url,
+            ))
+
+        except error.ConnectError as e:
```
Since we are construing url based on data contained in params , should we catch exception for the case when url is invalid? 
Similar to https://ews-build.webkit.org/#/builders/70/builds/23670/steps/17/logs/err_text
`builtins.ValueError: Invalid URI b'https://results.webkit.org/api/results-summary/layout-tests/gamepad/gamepad-polling-access.html?platform=mac&style=release&flavor=wk2&ref=256843.3@(HEAD detached at 1d7abcd180ab)'`

dewei-zhu commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -4926,25 +5003,32 @@ def evaluateCommand(self, cmd):
         try:
-            response = requests.get(self.url_for_hash_details(hash), timeout=60)
+            response = yield TwistedAdditions.request(
+                url=f'{COMMITS_INFO_URL}{hash}/json', logger=print,
```
Is `url_for_hash_details` used elsewhere since we just removed its usage here.

JonWBedard replied:
Good catch, removing.

dewei-zhu reviewed (approved):
r=me

JonWBedard committed:
[ews-build.webkit.org] Use twisted deferred to communicate with commits.webkit.org
https://bugs.webkit.org/show_bug.cgi?id=249174
rdar://103269096

Reviewed by Dewei Zhu and Aakash Jain.

* Tools/CISupport/ews-build/steps.py:
(DetermineLandedIdentifier): Inherit from ShellCommandNewStyle.
(DetermineLandedIdentifier.__init__): Ditto.
(DetermineLandedIdentifier.run): Replace start and evaluateCommand.
(DetermineLandedIdentifier.identifier_for_hash): Use TwistedAdditions.request and return
a deferred object.
(DetermineLandedIdentifier.start): Deleted.
(DetermineLandedIdentifier.url_for_hash_details): Deleted.
(DetermineLandedIdentifier.evaluateCommand): Deleted.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/258032@main
